fem/bilininteg_hcurl.cpp

```diff
@@ -366,6 +366,130 @@ void PAHcurlMassAssembleDiagonal3D(const int D1D,
    }); // end of element loop
 }
 
+void SmemPAHcurlMassAssembleDiagonal3D(const int D1D,
+                                       const int Q1D,
+                                       const int NE,
+                                       const Array<double> &_Bo,
+                                       const Array<double> &_Bc,
+                                       const Vector &_op,
+                                       Vector &_diag)
+{
+   constexpr static int VDIM = 3;
+   constexpr static int MAX_D1D = HCURL_MAX_D1D;
+   constexpr static int MAX_Q1D = HCURL_MAX_Q1D;
+
+   MFEM_VERIFY(D1D <= MAX_D1D, "Error: D1D > MAX_D1D");
+   MFEM_VERIFY(Q1D <= MAX_Q1D, "Error: Q1D > MAX_Q1D");
+
+   auto bo = Reshape(_Bo.Read(), Q1D, D1D-1);
```

Comment:
`const auto` might be also useful for the `Read`s - if the DeviceTensor evolves by doing these calls.

Reply:
This would need to be changed not just in this file but all the PA kernels (I don't see it being used in mass or diffusion kernels, for example). Since it would affect so many lines of code, I am hesitant to do this unless it is really necessary, which is unclear to me.

Reply:
sure - that ca be done later.